### ISSUE TITLE
Обработка "success" запроса

### ISSUE THREAD

**Issue by ange007** [hiqdev/omnipay-interkassa#5]
Приветствую.

На сколько я понимаю - **success** запрос после прохождения оплаты - не является запросом подтверждения, и следует ждать запроса по **notify** ссылке.

Но при текущей реализации - попытка обработать данные пришедшие с **Interkassa** _(что-бы получить например orderID)_ - вызывает исключение: `Undefined index: ik_sign` 
Что логично так как эти данные прийдут только в **notify** запросе.

Как в таком случае "универсально" считывать **orderID** средствами **OmniPay**?
Проверял это всё в тестовом режиме оплаты если что.

**Comment by hiqsol** [comment]
Да, **success** не надо слать на CompletePurchase, будет ошибка.

Мы реализовали только Purchase и CompletePurchase (**notify**) запросы.
Вполне можно реализовать запрос Success. Для этого надо закодить два класса:

- src/Message/SuccessRequest.php
- src/Message/SuccessResponse.php

Нам эта возможность не нужна, поэтому, сорри, мы не будем это делать.
Но сделать это не сложно, с удовольствием примем PR  :)

**Comment by ange007** [comment]
Если я верно понимаю логику **omnipay** - то библиотека рассчитана только на **authorize/completeAuthorize** и **purchase/completePurchase**.

Получается для этого нужно реализовывать часть с **authorize/completeAuthorize** ?
Я просто пока не совсем разобрался этой либой и со структурой _(одни шлюзы используют одно, другие другое)._

**Comment by hiqsol** [comment]
Насколько я понял логику omnipay - можно делать и свои, произвольные типы запросов.
У разных шлюзов разные возможности соответственно разные запросы, только purchase/completePurchase -  более менее есть у всех.

**Comment by ange007** [comment]
Ок, а как вы посмотрите если реализовать это немного по другому.
В случае если нет параметра **ik_sign** но есть **ik_co_id** и **ik_pm_no** - просто считывать переданные данные, без вывода исключения и установки **success = true**?

**Comment by hiqsol** [comment]
Нет, имхо логика omnipay такая что нужен отдельный request/response под это.
А чтоб не было много копипасты надо подумать как организовать общий функционал для success и completePurchase.

**Comment by ange007** [comment]
Исходя из их же примеров, всё в принципе должно сводится к:
1. Создали шлюз
2. Передали данные в **purchase/completePurchase** или **authorize/completeAuthorize**
3. Обработали ответ через: **isSuccessful**/**isRedirect**/**getMessage**/**getTransactionReference**

Просто добавлять какую-то специфическую функцию - это при получении подтверждения оплаты городить список `if( $service = * )` в случае использования нескольких провайдеров.

**Comment by ange007** [comment]
Или речь идёт о том что так оно и будет, просто нужно написать **request/response** которые будут  обрабатываться за фасадом **completePurchase** в случае условий которые я писал выше?

**Comment by hiqsol** [comment]
Как я это понимаю:
После оплаты от платёжная система выполняет два действия:
- отправляет подтверждение оплаты на notify url и это должно обрабатываться completePurchase request/response
- отправляет пользователя на success url и это должно обрабатываться success request/response

имхо неразумно пытаться использовать и там и там completePurchase

**Comment by ange007** [comment]
Так а что в таком случае использовать?
По сути то в **completePurchase** используется ведь и для проверки на "отмену" платежа _(если он не прошёл)_.

**Comment by hiqsol** [comment]
> Так а что в таком случае использовать?

в каком "в таком" ?

> По сути то в completePurchase используется ведь и для проверки на "отмену"

главное чтоб он использовался для проверки уведомления о платеже, а не тогда когда юзер возвращается с платёжной системы

**Comment by ange007** [comment]
> в каком "в таком" ?

В случае возврата с платёжной системы.
Но чтоб не потребовалось городить if`ы при работе с данным провайдером.

В https://github.com/collizo4sky/omnipay-2checkout/blob/master/src/Gateway.php вообще используется: 
`public function completePurchase(array $parameters = array())`
`public function acceptNotification(array $parameters = array())`
Хотя тут тоже используется доп. функция, но хотя-бы **acceptNotification** это обработка уведомление, а возврат на сайт - это **completePurchase** что более логично.

**Comment by ange007** [comment]
https://github.com/thephpleague/omnipay-common/blob/master/src/Omnipay/Common/AbstractGateway.php#L255:
`public function supportsAcceptNotification()`

Исходя из данной логики верно таки для уведомлений использовать: **acceptNotification**

**Comment by hiqsol** [comment]
Сорри, тут не помогу, про acceptNotification первый раз слышу.

**Comment by ange007** [comment]
Приветствую ещё раз - сделал форк _(правда не напрямую так как была бы путаница с кодовой базой)_.

Реализовал **acceptNotification**, **isPending** и прочие плюшки поддерживаемые **omnipay** для нормальной работы с библиотекой.
Правда пока убрал тесты _(так как нужно их написать ещё)_ и документацию _(в ней всё-равно ничего не было по сути)_.

https://github.com/ange007/omnipay-interkassa

**Comment by goodmind** [comment]
@hiqsol есть какой-нибудь прогресс? библиотека жива вообще?

**Comment by hiqsol** [comment]
Библиотека жива, мы используем ее в проде.

Прогресс чего интересует?
Был бы пул реквест мы могли бы его принять.

**Comment by ange007** [comment]
Я увы **PR** делать не смогу, так как те кто обновит библиотеку просто напросто сломают всё что у них до этого работало. Так как логика текущей версии отличается от "правильной" - которая диктуется архитектурой библиотеки **"OmniPay"** _(вывод сделан после изучения исходников библиотеки и непосредственно примеров от разработчиков OmniPay, правда почему-то многие не придерживаются её)_. Поэтому я и сделал для себя форк.

**Comment by cf-git** [comment]
Добрый день! Спасибо, вы создали отличный продукт. 
Возможно мой коммент сильно запоздал, но на всякий случай...
Сама интеркасса не отдает Сигн. Возможно, есть опции которые нужно включить. А поскольку она не отдает сигн, то проверку можно либо не проводить, либо проводить проверку хотябы IP, что само по себе бесполезно.... Кто столкнулся с подобной проблемой, можете не фокусироваться на решении, только трата времени. Комплит проводить не нужно.

**Comment by SilverFire** [comment]
Глянул логи, во свежий нотификейшн:

```
  array (
    'ik_co_id' => '***',
    'ik_co_prs_id' => '***',
    'ik_inv_id' => '***',
    'ik_inv_st' => 'success',
    'ik_inv_crt' => '2019-07-01 12:29:19',
    'ik_inv_prc' => '2019-07-01 12:30:51',
    'ik_trn_id' => '***',
    'ik_pm_no' => '***',
    'ik_pw_via' => 'mastercard_cpaytrz_merchant_rub',
    'ik_am' => '11.75',
    'ik_co_rfn' => '11.7500',
    'ik_ps_price' => '824.45',
    'ik_cur' => 'USD',
    'ik_desc' => '***',
    'ik_sign' => '8d17vrIj6AdYj537hkutYg==',
  ),
```

**Comment by cf-git** [comment]
Да. А сацес лог есть?

**Comment by cf-git** [comment]
Странно у меня нету сигна... Хотя я проверяю на более поздних стадиях. Меня пока интересует конечный результат.

```
local.NOTICE: Payment complete of interkassa, Order: 22 {"ik_co_id":"***","ik_inv_id":"***","ik_inv_st":"waitAccept","ik_pm_no":"***"}
```
11:34
```
local.NOTICE: Notification of interkassa, Order: 24 {"ik_co_id":"***","ik_co_prs_id":"***"","ik_inv_id":"***"","ik_inv_st":"success","ik_inv_crt":"2019-07-02 11:31:31","ik_inv_prc":"2019-07-02 11:31:31","ik_trn_id":***","ik_pm_no":"***"","ik_pw_via":"test_interkassa_test_xts","ik_am":"30","ik_co_rfn":"30","ik_ps_price":"30.9","ik_cur":"UAH","ik_desc":"***","ik_sign":"***""}
```

Можно попробовать проводить последний шаг... Но я не вижу в этом необходимости. Сервера интеркассы не будут достукиваться до сервера и ждать подтверждения того что вы получили уведомления... По крайней мере в дока апи об этом, я, ничего не нашел.
